### DISCUSSION TITLE
Visiting the interior of a shape

### DISCUSSION THREAD

**Discussion by Captain-Chaos** [locationtech/jts#928] [Q&A] [answered]
Hi. I'm the author of an interactive Swing-based program for creating Minecraft maps by painting 2D worlds. I'm looking for a geometry library to help me implement:

- Allowing the user to construct editable lines, splines, etc., with control points that can be dragged, etc.
- Painting shapes on the 2D world, iterating over boundaries or interiors of shapes so that I can apply some operation to the map at every point that lies inside the shape, or along a path

Would JTS be a good candidate for this? And more specifically: what would be the most efficient way with JTS to either:

- Visit all points along a path in an integer coordinate space
- Visit all points inside a shape in an integer coordinate space

**Comment by dr-jts** [comment, marked as answer]
If I'm understanding your requirements correctly, then JTS may or may not have objects and functions that do what you want.  

* You can create line and polygon geometry with integer coordinates.   There is no way to represent splines directly.  
* You could compute a grid of integer coordinates and then use JTS to test whether each one lies inside a polygon in an efficient way
* Not sure what you mean by "visit points along a path in integer coordinate space".  You could potentially do this by creating the grid of all points in the extent (bounding box) of a path (as a JTS LineString), and then test to find ones which are near the actual path (given some tolerance distance).

**Comment by Captain-Chaos** [comment]
> Not sure what you mean by "visit points along a path in integer coordinate space".

One of the main challenges I have is creating neat one pixel wide lines, that follow a path of a mix of straight lines and curves. "Neat" meaning that the pixels only adjoin at corners or opposite edges, so for example:

![image](https://user-images.githubusercontent.com/4256994/210363315-a070926a-b0bd-47c6-ba83-0bede2272e52.png)

And not:

![image](https://user-images.githubusercontent.com/4256994/210363337-e226f400-8a77-4634-966b-6f7edaf7ef2b.png)

I could probably come up with something myself that could do that, but I'm looking for a library which can do that kind of thing for me.

**Reply by dr-jts** [reply]
You might try looking at [**Bresenham's Line Algorithm**](https://en.wikipedia.org/wiki/Bresenham%27s_line_algorithm).  There should be Java implementations available.

And for the polygons perhaps you are looking for a [**flood-fill algorithm**](https://en.wikipedia.org/wiki/Flood_fill) - where again I expect there are Java implementations available.

**Reply by Captain-Chaos** [reply]
Thanks for the pointers! I think I hear you say that JTS might not be a good match for my requirements then.

**Reply by dr-jts** [reply]
> JTS might not be a good match for my requirements then.

Right, it sounds like that's the case.
